neo/Consensus/ConsensusService.cs

```diff
@@ -287,19 +333,154 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
             }
         }
 
+        private bool CheckPrimaryPayloadSignature(ConsensusPayload payload, PrepareRequest message)
+        {
+            /// <summary>
+            /// The Speaker Signed the Payload without any signature (this was the trick/magic part), PrepReqSignature was empty
+            /// But the payload was latter modified with his signature, 
+            /// We mean, the PrepareRequest message was filled with the PrepReqSignature of the Empty Payload and then serialized again into this Payload.
+            /// Thus, we need to remove the signature from the Payload to correctly verify Speaker identity agreements with this block
+            /// </summary>
+            message.PrepReqSignature = new byte[64];
+            payload.Data = message.ToArray();
+            if (!Crypto.Default.VerifySignature(payload.GetHashData(), context.SignedPayloads[payload.ValidatorIndex], context.Validators[payload.ValidatorIndex].EncodePoint(false)))
```

Comment:
We should use primary's signature from `message ` instead of `context.SignedPayload[payload.ValidatorIndex]`. It works fine when this method is called from `onPrepareRequestReceived()`. But when it's called from `onRenegeration()` after a CN reconnect to network, `context.SignedPayload` is empty. This throws Exception. 

It is the same for line 357.

Reply:
@wanglongfei88, thank for your attention on this. You are right.

It took us some debugging time for these lines because the message was also losing its signature when the assignment `message.PrepReqSignature = new byte[64];`.

 tempSignature was now created now: 
```cs
        private bool CheckPrimaryPayloadSignature(ConsensusPayload payload)
        {
            // TODO Maybe include some verification here
            PrepareRequest message = GetPrepareRequestMessage(payload);

            /// <summary>
            /// The Speaker Signed the Payload without any signature (this was the trick/magic part), PrepReqSignature was empty
            /// But the payload was latter modified with his signature, 
            /// We mean, the PrepareRequest message was filled with the PrepReqSignature of the Empty Payload and then serialized again into this Payload.
            /// Thus, we need to remove the signature from the Payload to correctly verify Speaker identity agreements with this block
            /// </summary>
            byte[] tempSignature = message.PrepReqSignature;
            message.PrepReqSignature = new byte[64];
            payload.Data = message.ToArray();
            if (!Crypto.Default.VerifySignature(payload.GetHashData(), tempSignature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
                return true;

            /// <summary>
            /// These next 2 lines could be removed, because payload is not anymore used
            /// it was already saved before changed in the context.PreparePayload... However, let keep things clean for now
            /// </summary>
            message.PrepReqSignature = tempSignature;
            payload.Data = message.ToArray();
            return false;
        }
```

PS: after everything is working as intended we come to code clean.
Some points are duplicated now but can be reduced after achievement of an table version.